services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,99 @@
+const {Provider} = require('./provider');
+const aws = require('aws-sdk');
+
+class AwsProvider extends Provider {
+  constructor({workerPool}) {
+    super({workerPool});
+    this.configSchema = 'config-aws';
+
+    const {region} = workerPool.config;
+    const {apiVersion} = workerPool.config;
+
+    aws.config.update({region});
+    this.iam = new aws.IAM({apiVersion});
+  }
+
+  async setup() {
+    // create IAM user
+    const {region} = this.workerPool.config;
+    const {account} = this.workerPool.providerData;
+    const userName = `wm-aws-provider-${this.workerPool.providerData.userName}`;
+
+    await readModifySet({
+      read: async () => await this.iam.getUser({UserName: userName}),
+      set: async () => await this.iam.createUser({UserName: userName}),
+    });
+
+    // create IAM policy
+    const ec2policy = {
+      "Statement": [{
+        "Effect": "Allow",
+        "Action": [
+          "ec2:DescribeAvailabilityZones",
+          "ec2:DescribeInstances",
+          "ec2:TerminateInstances",
```

Comment:
Those are just actions. The authorization itself is determined by the resources' ARN. At the moment, I am having `arn:aws:ec2:${region}:${account}:*` <---- which means every ec2 instance in that region and account. The star can be replaced with something more granular - suggestions?

and yeah, these permissions will apply to everything that will be able to use this user (so, probably every ec2 instance in that region).

> These seem like the permissions worker-manager itself will need rather than the permissions an instance will need.

I was understanding the `setup` method as provider setup (i.e. these would be permissions for the provider). So these are for workers then? Which permission would they need? Also, in that case, it looks like we would have to setup an `aws` component in the loader (in `main.js` for the provider)? That wasn't done for the gcp provider, so I was wondering

Reply:
> an instance

wait, do you mean "provider instance" or "ec2 instance"?

Reply:
Yeah, ec2 instance in this case. The service account that the google provider sets up isn't used by the google provider, it is used by the instances themselves for permissions such as writing to stackdriver. We set up an account for w-m itself to use with terraform and that's what we will do in the case of the aws provider as well I think.

Reply:
oh ok, git it.

My question then is this: which apis should aws workers be able to call? (Where will we log etc.)

Reply:
I wonder if we could rename the method to `setupWorkers`? I'll add a comment to the aws provider code in any case

Reply:
It can do more than setup workers. It is just the place to put anything you only want the provider to do once on startup. It doesn't go in the constructor because you may want to do async stuff here and you can't await that in the constructor. Just think of it as a second constructor I think.

Reply:
I think it doesn't do any worker setup, if by "worker setup" you mean starting workers.  I'm not sure what else it could mean :)

Reply:
> My question then is this: which apis should aws workers be able to call? (Where will we log etc.)

Probably the best way to answer this is to look at AWS provisioner.

Reply:
>  I'm not sure what else it could mean :)

for me, it means preparing things that workers will need - like service accounts or iam users

> Probably the best way to answer this is to look at AWS provisioner.

👍